### ISSUE TITLE
Enforce std::map and std::unordered_map value_type requirements

### ISSUE THREAD

**Issue by cjdb** [llvm/llvm-project#106635]
Per [[associative.reqmts.general]/8](https://eel.is/c++draft/associative.reqmts.general#8), the requirements placed on a container's `value_type` instead apply to its `key_type` and `mapped_type`. This means that the following should all be rejected by the compiler.

```cpp
std::map<int const, int> K1; // same as volatile
std::map<int, int const> M1; // same as volatile
std::map<int, int&> M2; // (int& as a key is already rejected for overload ambiguity reasons)
std::map<int&&, int> K3;
std::map<int, int&&> M3;
std::map<int, int[]> M4;
std::map<int[n], int> K5;
std::map<int, int[n]> M5;
```
```cpp
std::unordered_map<S const, int> K1; // (int const as a key is already rejected because std::hash<int const> isn't allowed)
std::unordered_map<int, int const> M1; // same as volatile
std::unordered_map<int, int&> K2;
std::unordered_map<int, int&&> K3;
std::unordered_map<int, int[]> M4;
std::unordered_map<int, int[2]> M5;
```

**Comment by cjdb** [comment]
I would like to deprecate support in LLVM 20, make it opt-in for LLVM 21, and remove support in LLVM 22.

**Comment by frederick-vs-ja** [comment]
In microsoft/STL#3660 @StephanTLavavej said that `mapped_type` is allowed to be a reference. Which is correct?

**Comment by StephanTLavavej** [comment]
To be clear, I think that references are wacky here, and I'd like to forbid them. We'd need to deal with the existing usage though.

**Comment by cjdb** [comment]
> We'd need to deal with the existing usage though.

I agree. We need to phase it out, rather than simply dropping support. I'm not particularly fixed on three versions to remove either: that's just an initial proposal.

**Comment by cjdb** [comment]
Closing this issue until there's committee guidance regarding my comment in cplusplus/draft#7249.

**Comment by cjdb** [comment]
Casey clarified that the requirements remain the same, and that `key_type` and `mapped_type` are both required to be [_Cpp17Erasable_](https://eel.is/c++draft/containers#container.alloc.reqmts-2.6). This means that all the restrictions we place on non-map containers' `value_type`s apply to map types' `key_type`s and `mapped_type`s.
